[WIP] support transparent & ctermbg cursor + minimal highlight api

teto opened the pull request:
See https://github.com/neovim/neovim/issues/3688 and https://github.com/neovim/neovim/issues/6428 .
- [x] support for a transparent TUI cursor (set guibg=None to the cursor highlight).
- [x] ctermbg can also affect cursor color even whithout termtruecolors.
- [x] notify UIs about changed highlights
- [x] add api to play with highlights: nvim_hl_from_id/name
- [x] simplify handling of the 'Normal' group. :hi Normal should work even with -u NORC.
- [ ] add to wiki a table describing terminal supported ? (for cursor color/transparency etc)
- [x] reset cursor for VTE (seems to be a VTE extension `printf "\x1b[0 q"`)

As for setting cursor foreground color, VTE api allows it (termite has a cursor_foreground parameter for instance) but there is no OSC yet. https://bugzilla.gnome.org/show_bug.cgi?id=695011 proposes OSC20 for that but  rxvt has a pre-existing use of OSC 20 (for background pixmap). Thomas Dickey (xterm maintainer) seems ok with a patch that would set cursor foreground color, so anyone feel free to propose such a patch.

teto committed:
tui: support gui=inverse and transparent colors

remove some unneeded toggling

fix most problems though in a hackish manner

brcolow commented on src/nvim/tui/tui.c:
```diff
@@ -576,11 +576,30 @@ static void tui_set_mode(UI *ui, ModeShape mode)
+
+    if (ui->rgb) {
+
+      RgbValue bg = aep->rgb_bg_color;
```
Is this more succinct using a ternary?

```c
// if gui=reverse, swap bg and fg colors
RgbValue bg = aep->rgb_ae_attr == HL_INVERSE ? aep->rgb_fg_color : aep->rgb_bg_color;
```
Just a random thought..might not be worth changing.

teto replied:
Ends up > 80 chars so it ends up not being clearer

brcolow commented on src/nvim/tui/tui.c:
```diff
@@ -913,12 +932,15 @@ static void out(void *ctx, const char *str, size_t len)
     const char *val)
 {
-  if (!unibi_get_str(ut, str)) {
+  const char *temp = unibi_get_str(ut, str);
```
Why use temp here? Can't one just:

```c
if (!unibi_get_str(ut, str)) {
  unibi_set_str(ut, str, val);
  return true;
}
return false;
```

teto replied:
Leftover of a debug session. Removed the temp variable. Kept returning the bool as it helps debug and should be optimized out by compiler anyway.

justinmk commented on src/nvim/tui/tui.c:
```diff
@@ -576,11 +576,30 @@ static void tui_set_mode(UI *ui, ModeShape mode)
+      RgbValue bg = aep->rgb_bg_color;
+
+      // if gui=reverse, swap bg and fg colors
+      if (aep->rgb_ae_attr == HL_INVERSE) {
```
should it be `aep->rgb_ae_attr & HL_INVERSE` ?

teto replied:
Fixed.

justinmk commented:
In https://github.com/neovim/neovim/issues/6425#issuecomment-291646189 @zhou13 mentioned control sequence `\x1b]112\x07` for "resetting to default cursor color (i.e., `gui=reverse`)".

teto commented:
@zhou13 I can't find out what `\x1b]112\x07` does ? In gvim gui=reverse swaps guifg and guibg. This PR changes the guibg of the cursor but not the guifg as it should (aka the text color underneath the cursor). Not sure how to properly do that on a terminal though.

I don't get why I have this error: terminal tests look ok for the rest :/

```
[  ERROR   ] 1 error, listed below:
[  ERROR   ] test/functional/terminal/tui_spec.lua @ 282: tui focus event handling can handle focus events in terminal mode
./test/functional/helpers.lua:252: 
retry() attempts: 3
./test/functional/ui/screen.lua:301: Row 1 did not match.
Expected:
  |*ready $                                           |
  |[Process exited 0]{1: }                               |
  |                                                  |
  |                                                  |
  |                                                  |
  |gained                                            |
  |{3:-- TERMINAL --}                                    |
Actual:
  |*nal                                               |
  |:terminal                                         |
  |{1: }                                                 |
  |{4:~                                                 }|
  |{4:~                                                 }|
  |gained                                            |
  |{3:-- TERMINAL --}                                    |
```
Thus I checked out terminals (I don't use the feature) and saw that in both master and this PR I get some (red) artifacts:
![2017-04-23-023141_764x956_scrot](https://cloud.githubusercontent.com/assets/886074/25309418/38fdd250-27cd-11e7-8717-2e973fc3cacf.png)
could that be linked ?

zhou13 commented:
````
echo -ne '\e]12;#ff00ff\a'  # your cursor becomes purple.
echo -ne '\e]112\a'  # reset your cursor
````
tested in xfce4-terminal.

justinmk commented:
> Thus I checked out terminals (I don't use the feature) and saw that in both master and this PR I get some (red) artifacts:

What terminal? I didn't see that on iTerm.

Is the test failing consistently?

teto commented:
Stupid me, it's just a plugin showing space at end of line. Thought it was linked to guicursor because I used to test with a red cursor.

The tests pass on master, but fail consistently for this PR (on travis too). Seems like       `feed_data(':terminal\n')` is not run as it appears in output. Will debug this later, I am clueless right now :'(

jamessan commented:
> I can't find out what \x1b]112\x07 does ?

It's a Operating System Command ([OSC]).  They're of the form `OSC PS ; PT BEL`.

Any terminal supporting these should respond to the PR.  xterm does, if `allowFontOps` is enabled.  Funnily, pangoterm doesn't yet.

> The 10 colors (below) which may be set or queried using 1 0
 through 1 9  are denoted dynamic colors, since the correspond-
 ing control sequences were the first means for setting xterm's
 colors dynamically, i.e., after it was started.  They are not
 the same as the ANSI colors.  These controls may be disabled
 using the allowColorOps resource.  At least one parameter is
 expected for Pt.  Each successive parameter changes the next
 color in the list.  The value of Ps tells the starting point
 in the list.  The colors are specified by name or RGB specifi-
 cation as per XParseColor.
 ...
 Ps = 1 0  -> Change VT100 text foreground color to Pt.
 Ps = 1 1  -> Change VT100 text background color to Pt.
 Ps = 1 2  -> Change text cursor color to Pt.
 Ps = 1 3  -> Change mouse foreground color to Pt.
 Ps = 1 4  -> Change mouse background color to Pt.
 ...
 The dynamic colors can also be reset to their default
 (resource) values:
 Ps = 1 1 0  -> Reset VT100 text foreground color.
 Ps = 1 1 1  -> Reset VT100 text background color.
 Ps = 1 1 2  -> Reset text cursor color.
 Ps = 1 1 3  -> Reset mouse foreground color.
 Ps = 1 1 4  -> Reset mouse background color.

[OSC]: http://invisible-island.net/xterm/ctlseqs/ctlseqs.html#h2-Operating-System-Commands

teto committed:
Fix tui_spec retrial

teto commented on src/nvim/api/private/helpers.c:
```diff
@@ -83,6 +84,105 @@ bool try_end(Error *err)
+      } else {
+        attrs.background = aep->cterm_bg_color - 1;
+      }
+      ILOG("attr=%d  color to %d", attr_code, attrs.background);
```
todo remove

teto commented on src/nvim/syntax.c:
```diff
@@ -8251,9 +8256,37 @@ color_name_table_T color_name_table[] = {
+  }
+
+  return NULL;
+}
```
shouldn't this function along with cterm_color_name and color_numbers_X live in one of the terminal libraries ?

justinmk replied:
It's not clear to me why we need this function. We only need the RGB form sent with mode_info_set.

justinmk replied:
Not that I know of.

teto replied:
I do not know the scope of libtermkey or unibilium but this part should apply to all terminals so it might make sense to have it there.

teto commented on src/nvim/syntax_defs.h:
```diff
@@ -67,9 +67,9 @@ struct syn_state {
-  int16_t rgb_ae_attr, cterm_ae_attr;  // HL_BOLD, etc.
+  int16_t rgb_ae_attr, cterm_ae_attr;  ///< \ref HL_ATTRIBUTES
   RgbValue rgb_fg_color, rgb_bg_color, rgb_sp_color;
   int cterm_fg_color, cterm_bg_color;
```
if the hl_group retains the original cterm values, cterm_Xg_color could now be RgbValues too ?!

teto commented on src/nvim/tui/tui.c:
```diff
@@ -91,13 +91,14 @@ typedef struct {
     int enable_bracketed_paste, disable_bracketed_paste;
     int set_rgb_foreground, set_rgb_background;
-    int set_cursor_color;
+    int set_cursor_bg_color;
```
renamed for clarity as some terms (iTerm?) seem to have a `set_cursor_fg_color`

teto commented on src/nvim/tui/tui.c:
```diff
@@ -482,6 +500,9 @@ static cursorentry_T decode_cursor_entry(Dictionary args)
       r.id = (int)value.data.integer;
+    } else if (strequal(key, "hl_dict")) {
+      r.attrs = decode_hl_entry(value.data.dictionary);
+      ILOG("decoded hl_dict with bg=%d", r.attrs.background);
```
remove

teto commented on src/nvim/tui/tui.c:
```diff
@@ -587,12 +608,21 @@ static void tui_set_mode(UI *ui, ModeShape mode)
+    if (cursor_bg == -1) {
+      unibi_out(ui, unibi_cursor_invisible);
+    } else {
+      unibi_out(ui, unibi_cursor_normal);  // display if previously invisible
```
this call could be prevented with an additionnal check if we save the visibility status. worth it ?

justinmk replied:
It only happens on mode-change, so it's ok. Incidentally we do unibi_cursor_normal quite often, it's part of UI "flush".

justinmk replied:
In fact this call may not be needed at all, the next UI flush will show the cursor again.

teto commented on src/nvim/ui.c:
```diff
@@ -388,10 +388,10 @@ void ui_cursor_goto(int new_row, int new_col)
-  api_free_array(style);
+  UI_CALL(mode_info_set, enabled, mode_style_array(ui->rgb));
+  // todo
+  // api_free_array(style);
```
with the new style I can't free the array.

justinmk replied:
Because `HlAttrs` isn't an `Object`? Or why?

teto reviewed (commented):
supporting the "gui=inverse" thing seems hard so I dropped it.
I haven't tested this much but cursor color seems to follow ctermbg with set notermguicolors.
It is possible to hide cursor with guibg=None.
guicursor is resent for each highlight_changed(). Less then optimal but cursor is updated properly.

The easiest way to deal with guicursor would be to send the full mode description on mode change (=stateless) but it was suggested that as this information doesn't change often, it can be stateful. Hence it is sent via mode_info_set. Sending guicursor colors in mode_change seems like a bad compromise between the 2 as cursor highlights shouldn't change much either.

UIs will need to retrieve highlights in all cases (because externalized components will need to match vim colors), so I believe letting the UIs subscribe to chosen highlight changes (such as the cursor's highlights) would be the best approach ? . 
Meanwhile as a quickfix, I sent cursor colors along with guicursor info.
If this is blocking for 0.2 feel free to take over as I am not sure when I can complete this.

justinmk commented on src/nvim/syntax.c:
```diff
@@ -257,6 +255,48 @@ static int current_flags = 0;
+  "Cyan", "LightCyan", "Red", "LightRed", "Magenta",
+  "LightMagenta", "Yellow", "LightYellow", "White", "NONE"
+};
+// terminals with less than 16 colors...
```
We don't care about them for 'guicursor' purposes.

teto replied:
@mhinz asked to be able to change cursor via setting ctermbg. ctermbg=ColorName is straightforward but `ctermbg=4` needed a function to convert the integer to the colorname.

justinmk commented on src/nvim/tui/tui.c:
```diff
@@ -1060,7 +1093,7 @@ static void flush_buf(UI *ui, bool toggle_cursor)
   TUIData *data = ui->data;
 
-  if (toggle_cursor && !data->busy) {
+  if (toggle_cursor && !data->busy && cursor_bg != -1) {
```
Would be better to somehow communicate this via the existing `toggle_cursor`.

justinmk commented on src/nvim/tui/tui.c:
```diff
@@ -1021,10 +1053,11 @@ static void fix_terminfo(TUIData *data)
         ut, NULL, "\033]Pl%p1%06x\033\\");
   } else {
-    data->unibi_ext.set_cursor_color = (int)unibi_add_ext_str(
+    // new form rgb:X/X/X does not seem to work so well
```
outdated comment?

teto replied:
It's up to date but might be more relevant in the git log. rgb:X/X/X is supposed to work (see xparsecolor) but there are different forms such as rgb:XXX/XXX/XXX, rgb:XX/XX/XX and not all had an effect in termite. I would have to ask on vte tracker.

jamessan replied:
vte is [supposed](https://git.gnome.org/browse/vte/tree/src/vtetypes.cc?id=a31a41d2a93ef45c3940d714f82f58c6b3f9d7d8#n84) to convert the `rgb:rrr/ggg/bbb` (independent of the arity of each color) to the `#rrrgggbbb` form and then pass that on to [`pango_color_parse`](https://developer.gnome.org/pango/stable/pango-Text-Attributes.html#pango-color-parse).

Older versions of vte used [`gdk_parse_color`](https://developer.gnome.org/gdk3/stable/gdk3-Colors.html#gdk-color-parse) instead, but either one should handle the format fine.

If you're seeing different behavior, then I'd either guess it's an old version of vte or the terminal using vte is doing something wrong.

teto committed:
works so far. Whenever  a group is the first checked, Normal is inserted so that it becomes the first

teto commented:
code might be hard to follow as there is some dead code.
The idea was to have UIs register the highlight (=hl) groups (via `nvim_ui_subscribe_to_highlights`) for which they want to get notifications upon change. These subscriptions could be registered in `ui->subscribed_highlights`.
Patch tracks updated highlights in the global value (was easier for tests but it doesn't need to be global) `changed_highlights`. Then when `highlight_changed` is called, `ui_notify_changed_highlights` would send the updated `HlAttr`. In the end, I did not filter updates based on `ui->subscribed_highlights` (hence the dead code) as it seemed a bit overkill, there are not so many updates after startup.
In the TUI code, `tui_event` checks if the updaded highlights can change the cursor appearance (see `map_hlid_to_mode`) in which case it refreshes the cursor.

A few interrogations:
- having each UI check if the new highlight triggers a cursor update seems a waste. There could be a field hinting that cursor might need to be redrawn, or a specialization of refresh/flush for cursor.
- how to get highlights at startup ? current value could be sent right upon subcriptions `nvim_ui_subscribe_to_highlights` (once the UI gets guicursor value, it knows which highlights to subscribe to).
- I am not sure the current code tracks all highlight changes. Sounds ok when changing on command line. A (better) API to deal with hl might help there.

teto committed:
Normal group is inserted in first position

Removes the need for the +/-1 id dance when using HL_TABLE().

teto committed:
Fixed linting + test failure of using wrong hl id

teto committed:
Fixed handling of hl link + a few tests. yet there are more to fix

teto committed:
Fixes match_function_spec.lua test

teto committed:
All tests now passing.

Need to check id is correct for UserX highlights and also syntax.c:3670.
Also need to actually enable the TUI cursor functionalities for now
commented out + fixing terminal capabilities info.

teto committed:
Reestablished ui_bridge_event

teto commented:
transparent cursor works in termite. cursor color can also now be set via ctermbg + notermguicolors. I've added tests for the highlight api. Also nvim -u NORC -c ":hi Normal" now works as it should. 
There are many things in the PR so I might polish a few things and split it into separate PRs.

bfredl commented on src/nvim/api/vim.c:
```diff
@@ -54,6 +54,92 @@ void nvim_command(String command, Error *err)
+///
+/// @param highlights a list of highlights id (string or integer) to look for
+/// todo return array instead ? ArrayOf(HlAttrs)
+Object nvim_hl_get_list(Array highlights, Error *err)
```
if this is a batched version of the methods above then it is not needed (use call_atomic or lua)

bfredl commented on src/nvim/tui/tui.c:
```diff
@@ -805,6 +841,16 @@ static void tui_set_icon(UI *ui, String icon)
 // to make a copy for the tui thread
 static void tui_event(UI *ui, char *name, Array args, bool *args_consumed)
```
this shouldn't be used. In fact ui_event impl should be just deleted from ui_bridge and tui. (the infrastructure checks for funptr being NULL, so this is safe). Instead, mark events as not REMOTE_ONLY to have them acessible from TUI.

bfredl commented on src/nvim/ui.c:
```diff
@@ -182,6 +183,37 @@ static void ui_refresh_event(void **argv)
+  if (refresh_cursor) {
+    ILOG("updating cursor !");
+    Array temp = ARRAY_DICT_INIT;
+    ui_event("refresh_cursor", temp);
```
Changing highlights should be rare enough for UI:s to afford to update cursor on every highlight update event, alternative the ui could calculate the check themselves.

bfredl replied:
alternatively, just send another ui_cursor_shape(), no need for a separate event

teto replied:
ui_cursor_shape() tries to be smart and doesn't send the event if the mode hasn't changed since last update. I would need to add a "force_update" boolean to bypass that behavior. Would that be ok ?

teto replied:
Nevermind, as you say UIs can redraw cursor on every update. That's what I went for.

bfredl replied:
That is fine, but a `force` flag wouldn't be unprecedented either.

bfredl commented on src/nvim/ui_bridge.c:
```diff
@@ -180,3 +182,24 @@ static void ui_bridge_suspend_event(void **argv)
   ui->suspend(ui);
 }
+
+static void ui_bridge_event(UI *b, char *name, Array args, bool *args_consumed)
```
remove, add the relevant declations to ui_events.in.h and autogeneration takes care of this.

bfredl commented on src/nvim/ui.c:
```diff
@@ -182,6 +183,37 @@ static void ui_refresh_event(void **argv)
+
+  if (args.size) {
+    ILOG("submit %d updated hl", args.size);
+    ui_event("highlights", (args));
```
ui events aren't varadic, this should be a nested array. But changing to `ui_call_highlights_changed(args)` will take care of this.

teto replied:
ok done

teto commented on src/nvim/syntax.c:
```diff
@@ -44,6 +64,10 @@
+static int invalid_group_id = -1;
+
+// TODO(teto): use array instead GA_EMPTY_INIT_VALUE ?
+static garray_T changed_highlights  = { 0, 0, sizeof(int), 10, NULL };
```
I converted the garray into an Array to remove the conversion needed when sending the event.

teto committed:
fix after review

teto commented:
I have a question on how to enable the transparency for the cursor on neovim.

As written in https://github.com/neovim/neovim/issues/6428, vim enables transparency when :
1/ the mapped hl (="highlight") for the current mode has no bg set (e.g., `:hi Cursor ctermbg=NONE guibg=NONE` hides the cursor).
2/ the highlight does not exist

Vim's default for guicursor is:
```
  guicursor=n-v-c:block-Cursor/lCursor,ve:ver35-Cursor,o:hor50-Cursor,i-ci:ver25-Cursor/lCursor,r-cr:hor20-Cursor/lCursor,sm:block-Cursor-blinkwait175-blinkoff150-blinkon175
```
while neovim's is:
`guicursor=n-v-c-sm:block,i-ci-ve:ver25,r-cr-o:hor20`

When no highlight (or a wrong hl) is mapped, the TUI doesn't touch the color to prevent using ansi color codes that could mess with terminals. It means neovim can't use 2/ to hide the cursor. Remaining solution is 1/ i.e., detect that ctermbg=NONE or guibg=NONE. I can do that for the TUI in the rgb case, see syntax.c:
```
          if (STRCMP(arg, "NONE")) {
            HL_TABLE()[idx].sg_rgb_fg_name = (uint8_t *)xstrdup((char *)arg);
            HL_TABLE()[idx].sg_rgb_fg = name_to_color(arg);
          } else {
            HL_TABLE()[idx].sg_rgb_fg_name = NULL;
            HL_TABLE()[idx].sg_rgb_fg = -1;
          }
```
the TUI can check if `HL_TABLE()[cursor_hl_id].sg_rgb_fg_name = NULL;`but in the generic case where UIs only have intel through the  HlAttr.background integer, it's different because setting the Cursor guibg to NONE will set the HlAttr.background value to the 'Normal' hl one. Thus if the 'Normal' bg is set the color of the cursor will be the one of the Normal hl instead of being transparent.
Another problem is that when running `make && ./build/bin/nvim -u NORC -c "set termguicolors" -c "set guicursor=n-v-c:block-Cursor/lCursor,ve:ver35-Cursor,o:hor50-Cursor,i-ci:ver25-Cursor/lCursor,r-cr:hor20-Cursor/lCursor"` with the default highlights, you have  `Cursor guibg=fg`i.e. cursor will use the 'Normal' bg but by default the normal bg is uninitialized so the cursor becomes transparent.

The previous paragraph was to add some context but in the end the possible solutions I have in mind are:
1/ use a different value for a NONE color than for an unset guibg. (-2 and -1 for instance)
2/ use a cursor_on/cursor_off ui event as it already exists for the mouse (and these functions seem to exist as well in vim).


Nb: gvim acts weird if you `set guicursor=n:block`, i.e., no hl selected, than do `:hi Cursor guibg=blue` it will change color whereas the cursor should not be influenced anymore.

teto committed:
Now bridges ui_bridge_highlights_changed events

justinmk commented on runtime/doc/options.txt:
```diff
@@ -2828,6 +2828,17 @@ A jump table for the options with a short description can be found at |Q_op|.
+vte-based (> 3.9) |  x  |  x  |      x     |
+iterm             |  x  |  x  |            |
+Konsole           |  x  |     |            |
+Rxvt              |  x  |     |            |
```
Would prefer not to have terminal minutiae in the docs. It could live in the FAQ on the wiki (and perhaps `man nvim`), but in general users should understand that Nvim does the best it can, and if it doesn't work in their terminal then it's just not supported.

justinmk commented on src/nvim/api/private/defs.h:
```diff
@@ -129,5 +129,9 @@ struct key_value_pair {
+typedef struct {
+  bool bold, underline, undercurl, italic, reverse;
+  int foreground, background, special;
+} HlAttrs;
```
why is this in API? Perhaps move it to globals.h if the old location caused a cycle. Or `ui_defs.h`.

justinmk commented on src/nvim/api/private/helpers.c:
```diff
@@ -84,6 +85,103 @@ bool try_end(Error *err)
   return ERROR_SET(err);
 }
 
+bool attr2hlattr(int attr_code, bool use_rgb, HlAttrs *out)
```
helpers.c is turning into misc1.c... Can't these functions live in some more appropriate module?

teto replied:
src/ui.c then ? it allows to restore the original HlAttr definition.

justinmk replied:
sure, though I assumed you moved things away from ui.c for a reason.

justinmk commented on src/nvim/api/vim.c:
```diff
@@ -54,6 +54,58 @@ void nvim_command(String command, Error *err)
+/// @param name Highlight group name
+/// @return a highlight description in dictionary {'hl_id': X, 'bg': Y, 'fg': Z}
+/// @see nvim_hl_from_id
+Dictionary nvim_hl_from_name(String name, Error *err)
```
following `:h dev-api`, needs a verb. `nvim_get_hl_by_name`

teto replied:
nvim_{thing}_{action}_{arbitrary-qualifiers} so nvim_hl_get_by_name instead. Didn't know dev-api. Great.

bfredl replied:
`nvim_{thing}` should be preferably reserved for rpc objects like buffer. `nvim_get_hl_by_name` is better.

justinmk replied:
@teto I prefer namespacing things like you suggest, but as @bfredl mentioned, for global functions we are already stuck with `nvim_{action}`. `nvim_{thing}_{action}` is for "objects".

justinmk commented on src/nvim/api/vim.c:
```diff
@@ -54,6 +54,58 @@ void nvim_command(String command, Error *err)
+/// @param name Highlight group name
+/// @return a highlight description in a dictionary
+/// @see nvim_hl_from_id
+Dictionary nvim_hl_from_id(Integer hl_id, Error *err)
```
`nvim_get_hl_by_id`

teto replied:
nvim_hl_get_by_id

justinmk commented on src/nvim/syntax.c:
```diff
@@ -6137,9 +6207,10 @@ do_highlight(
-    for (int i = 1; i <= highlight_ga.ga_len && !got_int; ++i)
-      /* TODO: only call when the group has attributes set */
+    for (int i = 0; i < highlight_ga.ga_len && !got_int; i++) {
+      // todo: only call when the group has attributes set
```
`TODO(vim)`

justinmk commented on src/nvim/syntax.c:
```diff
@@ -42,8 +61,14 @@
+#include "nvim/api/private/helpers.h"
 
 static bool did_syntax_onoff = false;
+static int invalid_group_id = -1;
```
should be either a constant named `kFoo` or a define `FOO`.

teto replied:
I renamed it to kInvalidGroupId.

justinmk commented on src/nvim/syntax.c:
```diff
@@ -7227,61 +7297,73 @@ static int syn_add_group(char_u *name)
-/*
- * Translate a group ID to the final group ID (following links).
- */
+bool hl_invalid_id(int hl_id)
```
FUNC_ATTR_PURE. can it also be static?

same for `hl_is_valid()`.

teto replied:
I added the FUNC_ATTR_PURE. I renamed hl_is_valid to hl_valid_id. I think it makes sense to keep one function available (either hl_invalid_id or hl_valid_id) for other units for instance it is used in tui.c. I can inline or remove the 2nd one.

justinmk commented on src/nvim/api/ui_events.in.h:
```diff
@@ -58,6 +58,8 @@ void set_title(String title)
     FUNC_API_SINCE(3);
 void set_icon(String icon)
     FUNC_API_SINCE(3);
+void highlights_changed(Array highlights_changed)
```
The naming precedent is "set", we don't have "changed" (though we do have "change"... oh dear) and the distinction isn't meaningful in this case. We already have `highlight_set`, which is a UI change, whereas this new event is a property change. 

Kind of a mess. Naming this `highlight_info_set` at least is analogous to `mode_info_set`. We can use "info" as the pattern for "property changes".

bfredl replied:
Indeed "set" is better. The UI event stream tells the current state, not whether something changed or not

justinmk commented:
> 1/ use a different value for a NONE color and for an unset guibg (right now, both use -1 making the distinction impossible between the two. I propose to use -2 for NONE and -1 for unset)
> 2/ use a cursor_on/cursor_off ui event as it already exists for the mouse (and these functions seem to exist as well in vim).

@teto No objection to 1/, but 2/ sounds like the UI is making decisions to work around TUI limits. 

In general I suggest: give UIs all relevant info and let the UIs make the decisions. If TUI sees that guifg=bg it can decide to hide the cursor. It does that by inspecting `highlight_info_set`.

teto committed:
reviews

bfredl commented:
> If TUI sees that guifg=bg it can decide to hide the cursor.

I think it is better to encode explicitly that the cursor should be hidden, one would expect this to happen consistently in both TUI and GUI, for instance by using -2 like option 1.

teto commented:
I went with option 1 as it was just a few lines of code. Setting guibg=NONE now translates to -2 in the datastructure instead of -1.  The PR now seems to behave as intended. I'll clean up a bit, try to split it into 2 PRs as the changes to hl group index is not related and might be seen as potentially harmful.

justinmk commented on runtime/doc/options.txt:
```diff
@@ -2803,7 +2803,9 @@ A jump table for the options with a short description can be found at |Q_op|.
-			for the cursor
+			for the cursor. If an invalid group is set, the TUI won't change colors
+      contrary to vim. If the group has guibg=NONE then the cursor becomes
+      invisible.
```
did you mean UI instead of TUI? (We can't control what GUIs do, but we can add to the guidelines at `:help dev-ui` which should be considered the "official UI behavior").

Can elide "contrary to vim".

justinmk commented on src/nvim/tui/tui.c:
```diff
@@ -474,7 +474,7 @@ CursorShape tui_cursor_decode_shape(const char *shape_str)
 static bool cursor_visible(void)
 {
-  return cursor_bg != -1;
+  return cursor_bg != -2;
```
kColorFoo ?

justinmk commented on src/nvim/ui.c:
```diff
@@ -142,6 +142,104 @@ void ui_event(char *name, Array args)
   }
 }
 
+bool attr2hlattr(int attr_code, bool use_rgb, HlAttrs *out)
```
sorry, I should have clarified. These functions should live in syntax.c (wasn't that where they originally were?)

jdebp commented on src/nvim/tui/tui.c:
```diff
@@ -933,12 +986,14 @@ static void out(void *ctx, const char *str, size_t len)
     unibi_set_str(ut, str, val);
+    return true;
   }
+  return false;
```
This is rather pointless, since nothing uses the return value; and for what it's worth I have something in the works that I hope will do away with quite a number of calls to this function.

yevhen-m commented:
Guys, so cool that you are working on this, thanks a lot!

teto commented:
Trying to write some tests to check for certain sequences such as `\x1b]0 q` IIRC. In some PR @justinmk you wrote:

> You can test terminal responses using a :terminal and hooking up a on_stdout handler. We have some tests doing that. This would at least verify that for a given $TERM we send/receive expected sequences.

Any precise test  in mind ? I've been lookling at job_spec.lua and do sthg like:


```
    helpers.source([[
    function! s:OnEvent(id, data, event) dict
      let userdata = get(self, 'user')
      let data     = a:data
      call rpcnotify(g:channel, a:event, userdata, data)
    endfunction
    let g:job_opts = {
    \ 'on_stdout': function('s:OnEvent'),
    \ 'on_stderr': function('s:OnEvent'),
    \ 'on_exit': function('s:OnEvent'),
    \ 'pty': 1,
    \ 'user': 0
    \ }
    ]])

-- feedkeys('itoto') should enter insert mode so that I can generate guicursor for insert mode and check for the presence of matching CSI
      nvim('command', "let j = jobstart([ '".. helpers.nvim_prog.."', '-u', 'NONE', '-i', 'NONE', '--cmd', 'set noswapfile','--cmd', 'call feedkeys(\"itoto\")' ], g:job_opts)")
    end

    local msg = next_msg()
      print(msg[3][0])
      local line = msg[3][2]
    -- here I would like to check if the cursor change sequence is present, with some regex for instance:
    -- line.find("]3 q\007toto")

    -- eq({'notification', 'stdout', {0, {'abc', ''}}}, )
```
my trouble is that the on_stdout seems to be called many times, not sure against which line to check or how many times I should call next_msg (if incorrect, it leaves the test pending).

teto committed:
Test starts looking good

teto reviewed (commented):
I am a bit at a loss as how to use jobsend to send input to a background nvim.
I would like for instance to insert "toto" and then go back to normal mode.
So I send `jobsend(j, "itoto\<c-c>:q!")` but the on_stdout callback isn't called for "toto", just for `<c-c>:q!`.
then I try sthg like `jobsend(j, "itoto\<c-r>\<c-l>:q!")`
and it gives me 2 lines of output or is it the same line redrawn ? why the \13 ?

```
{
  "notification",
  "onout",
  {
    {
      "itoto^R\13",
      "itoto^L:q!"
    }
  }
}
``` 

This is still WIP but to see the behavior, one can run
`LANG=C TEST_FILE=test/functional/terminal/tui_spec.lua make functionaltest 2>&1 >log && nvim log`

bfredl commented on src/nvim/api/vim.c:
```diff
@@ -704,12 +756,21 @@ void nvim_unsubscribe(uint64_t channel_id, String event)
 }
 
+/// Translate to integer if \p name is an hex value (e.g. #XXXXXX),
+/// else look into \p color_name_table to translate a color name to  its
```
color_name_table sounds like an internal datastructure. Also the descriptions of `nvim_get_color_map` sounds like they use two different color tables.

bfredl commented on src/nvim/po/sv.po:
```diff
@@ -5,7 +5,7 @@ msgid ""
 msgstr ""
 "Project-Id-Version: Vim 7.1\n"
 "Report-Msgid-Bugs-To: \n"
-"POT-Creation-Date: 2014-07-02 21:54+0200\n"
```
Normally we don't update po files in unrelated PRs?

teto replied:
mistake on my side.

chrisduerr commented:
Is this still being worked on @teto? I'm interested in `reset cursor for VTE (seems to be a VTE extension printf "\x1b[0 q")`. If this PR is not active anymore, maybe it's possible to pick out some of the features and implement them separately? I'd be happy to help getting the VTE cursor reset working.

teto commented:
@chrisduerr I promised some users to get the transparent cursor working but the lack of dynamic update when changing highlights was a pain so I wanted to solve it first in https://github.com/neovim/neovim/pull/8006. For many reasons, I can't work on neovim at the moment so please go ahead. I would 
be very happy to see the reset behavior as a VTE-based terminal user. The code has changed much since the PR so you might be better off starting from scratch. Here is some place where the reset occurs  https://github.com/neovim/neovim/pull/6566/files#diff-3ac88981a139ac0eee448a6bd19c7569R198

chrisduerr commented:
Thanks for the quick response. I'm not sure if I can get to it this weekend, but I'll definitely look into it when I've got some time and this PR is still open. :)

teto commented:
The PR is likely to stick for a few more months/years xD feel free to ask if you have any question.

teto commented:
this is very old and almost everything has been implemented apart from the transparent TUI cursor. see https://github.com/neovim/neovim/pull/11519 for that
